File: 프로그래머스/3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.java
class Solution {

    private int[][] rotate90(int[][] a) {
        int n = a.length;
        int[][] r = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                r[j][n - 1 - i] = a[i][j];
            }
        }
        return r;
    }

    private boolean isUnlocked(int[][] board, int pad, int lockSize) {
        for (int i = 0; i < lockSize; i++) {
            for (int j = 0; j < lockSize; j++) {
                if (board[i + pad][j + pad] != 1) return false;
            }
        }
        return true;
    }

    public boolean solution(int[][] key, int[][] lock) {
        int m = key.length;
        int n = lock.length;

        int pad = m - 1;
        int size = n + pad * 2;

        int[][] board = new int[size][size];

        // 중앙에 lock 배치
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i + pad][j + pad] = lock[i][j];
            }
        }

        int[][] curKey = key;

        for (int rot = 0; rot < 4; rot++) {
            for (int x = 0; x <= size - m; x++) {
                for (int y = 0; y <= size - m; y++) {

                    // key를 board에 더하기
                    for (int i = 0; i < m; i++) {
                        for (int j = 0; j < m; j++) {
                            board[x + i][y + j] += curKey[i][j];
                        }
                    }

                    if (isUnlocked(board, pad, n)) return true;

                    // 원복
                    for (int i = 0; i < m; i++) {
                        for (int j = 0; j < m; j++) {
                            board[x + i][y + j] -= curKey[i][j];
                        }
                    }
                }
            }
            curKey = rotate90(curKey);
        }

        return false;
    }
}

File: 프로그래머스/3/60059. 자물쇠와 열쇠/자물쇠와 열쇠.kt
class Solution {

    private fun rotate90(a: Array<IntArray>): Array<IntArray> {
        val n = a.size
        val r = Array(n) { IntArray(n) }
        for (i in 0 until n) {
            for (j in 0 until n) {
                r[j][n - 1 - i] = a[i][j]
            }
        }
        return r
    }

    private fun isUnlocked(board: Array<IntArray>, pad: Int, lockSize: Int): Boolean {
        for (i in 0 until lockSize) {
            for (j in 0 until lockSize) {
                if (board[i + pad][j + pad] != 1) return false
            }
        }
        return true
    }

    fun solution(key: Array<IntArray>, lock: Array<IntArray>): Boolean {
        val m = key.size
        val n = lock.size

        val pad = m - 1
        val size = n + pad * 2

        val board = Array(size) { IntArray(size) }

        // 중앙에 lock 배치
        for (i in 0 until n) {
            for (j in 0 until n) {
                board[i + pad][j + pad] = lock[i][j]
            }
        }

        var curKey = key

        repeat(4) {
            for (x in 0..(size - m)) {
                for (y in 0..(size - m)) {

                    // key를 board에 더하기
                    for (i in 0 until m) {
                        for (j in 0 until m) {
                            board[x + i][y + j] += curKey[i][j]
                        }
                    }

                    if (isUnlocked(board, pad, n)) return true

                    // 원복
                    for (i in 0 until m) {
                        for (j in 0 until m) {
                            board[x + i][y + j] -= curKey[i][j]
                        }
                    }
                }
            }
            curKey = rotate90(curKey)
        }

        return false
    }
}